### PR TITLE
Builds user sentiment survey (closes #10, closes #18).

### LINKED ISSUE #10: Build user sentiment survey.
Blocked by #8

### LINKED ISSUE #18: Assemble full data packet.
As mentioned in #13, a payload will be created by creating an object out of a series of measurement classes instantiated from a predictable set of data points, e.g.

```js
const payload = [
  TelemetryId,
  AnotherMeasurement
].reduce((accumulator, Measurement) => {
  let metric = new Measurement(...argsFromSomewhere);
  accumulator[metric.name] = metric.data();
}, {});
```

This should be merged with the results from the survey (#15) to create a data packet to send to various telemetry brokers (#16, #17).

### PR DESCRIPTION
**Pull request by chuckharmston**
A little bit preliminary due to #8 and #59, but it does implement the current working survey design:

screen shot 2017-02-01 at 6 51 12 pm

Pay special attention to the [new note in the README](https://github.com/chuckharmston/pulse/blob/da5789a57763b81a9068d6ee296c9cd79f2ca04f/README.md#survey) about testing the survey.

### CODE REVIEW

File: src/webextension/survey/components/app.jsx [hunk 1]
```diff
@@ -0,0 +1,33 @@
+import React from 'react';
+import { connect } from 'react-redux';
+import Uri from 'urijs';
+
+import Survey from './survey.jsx';
+
+class App extends React.Component {
+  // Read intiial form values from querystring, ensuring that the `sentiment`
```
Comment (pdehaan) [inline src/webextension/survey/components/app.jsx]: typo: "intiial"

File: src/webextension/survey/components/app.jsx [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+  // value is a number.
+  getInitialValues() {
+    const values = Object.assign({
+      details: ''
+    }, new Uri(window.location.search).search(true));
+    if ('sentiment' in values) {
+      values.sentiment = parseInt(values.sentiment);
```
Comment (pdehaan) [inline src/webextension/survey/components/app.jsx]: nit: you should probably pass a radix to `parseInt()` due to occasional unexpected behavior (although not like we're expecting leading zero input): http://eslint.org/docs/rules/radix

File: src/webextension/survey/components/fields/details.jsx [hunk 3]
```diff
@@ -0,0 +1,13 @@
+import React, { Component } from 'react'
```
Comment (pdehaan) [inline src/webextension/survey/components/fields/details.jsx]: uber nit: missing `;`. I should also make sure that we're running ESLint against our new .jsx friends.

File: src/webextension/survey/components/app.jsx (hunk 1 above)
Comment (chuckharmston) [reply src/webextension/survey/components/app.jsx]: ✅

File: src/webextension/survey/components/fields/details.jsx (hunk 3 above)
Comment (chuckharmston) [reply src/webextension/survey/components/fields/details.jsx]: ✅

Comment (chuckharmston) [reply src/webextension/survey/components/fields/details.jsx]: Good catch that the JSX files weren't being prettied or linted; I missed that entirely.

File: src/webextension/survey/components/app.jsx (hunk 2 above)
Comment (chuckharmston) [reply src/webextension/survey/components/app.jsx]: ✅

File: src/webextension/survey/components/fields/details.jsx (hunk 3 above)
Comment (chuckharmston) [reply src/webextension/survey/components/fields/details.jsx]: It should now be linting JSX files with the `plugin:react/recommended` rules preset.

File: src/webextension/lib/util.js [hunk 4]
```diff
@@ -8,3 +10,13 @@ export const makePageAction = tab => {
   browser.pageAction.show(tab.id);
   logger.log(`Created pageAction for ${tab.id}`);
 };
+
+// If being run in the WebExtension context, send a message to the SDK add-on.
+// Otherwise use the regular logger.
+export const sendMessage = (type, payload) => {
```
Comment (chuckharmston) [inline src/webextension/lib/util.js:14]: @rpl Paired with the [webpack dev server](https://github.com/chuckharmston/pulse/blob/b41c0de4d2985a63c5f9b6c261fa1d2b1b37fcef/README.md#survey), this is how I'm handling the unclear context. Not particularly inventive, but it gets the job done.

File: src/webextension/survey/components/app.jsx [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+    // Ensure that optional fields are included in the payload, even if empty.
+    const defaultValues = { details: '' };
+
+    // Generate a UUID, then send to the host SDK add-on so that it can
+    // associate a future submission with the correct tab and window.
+    defaultValues.id = uuid();
+    sendMessage('id', defaultValues.id);
+
+    // Get `sentiment` and `type` from the querystring.
+    var qs = new Uri(window.location.search).search(true);
```
Comment (pdehaan) [inline src/webextension/survey/components/app.jsx:26]: Not sure about your religious feelings re: `var` vs `let`/`const`.

If you prefer `let` and `const` exclusively, we can add the following rules to our ESLint config:

```yaml
    "no-var": "error",
    "prefer-const": "error",
```

---

But note that if we adopt `prefer-const`, we'll have those false positives in the for..of loops, so we may have to add some `// eslint-disable-next-line prefer-const` tags so we can use `for (let tab of tabs) {...}`.

File: src/webextension/lib/util.js [hunk 6]
```diff
@@ -1,3 +1,5 @@
+/* global browser, console */
```
Comment (pdehaan) [inline src/webextension/lib/util.js]: You could add an ESLint `env` config of "webextensions" to the .eslintrc file and then not need to define these globals in the header comments:

```json
  "env": {
    "browser": true,
    "es6": true,
    "node": true,
    "webextensions": true
  },
```

File: package.json [hunk 7]
```diff
@@ -34,9 +42,10 @@
     "url": "git+https://github.com/mozilla/pulse.git"
   },
   "scripts": {
-    "build": "webpack",
-    "watch": "webpack --watch",
+    "build": "NODE_ENV=dev webpack",
+    "watch": "NODE_ENV=dev webpack --watch",
     "prettier": "prettier --single-quote --write 'src/**/*.js' webpack.config.js",
```
Comment (pdehaan) [inline package.json]: Nit: Change glob to `src/**/*.{js,jsx}`

File: src/webextension/survey/components/fields/reason.jsx [hunk 8]
```diff
@@ -0,0 +1,80 @@
...
+
+Reason.propTypes = {
+  max: React.PropTypes.number,
+  min: React.PropTypes.number,
+  name: React.PropTypes.string.isRequired,
+  sentiment: React.PropTypes.number,
+  value: React.PropTypes.string.isRequired
+};
+
+class ReasonField extends Component {
```
Comment (pdehaan) [inline src/webextension/survey/components/fields/reason.jsx:40]: Not sure if there is a React best practice for having multiple classes per file, or if we should move this class into a reasonfield.jsx file.

Comment (lmorchard) [reply src/webextension/survey/components/fields/reason.jsx:40]: Eh, it's not exported, so I'd say they're "internal" classes for the exported public component. I'd use a rule of thumb that there should be 1 public component per file.

File: README.md [hunk 9]
```diff
@@ -48,6 +48,17 @@ While the WebExtension:
 - Serves the survey to collect information from the user.
 
 
+### Survey
+
+Because extensions cannot run content scripts on other extensions' hosted pages, you cannot use the [React Developer Tools](https://addons.mozilla.org/firefox/addon/react-devtools/) to debug the survey in context of the add-on. To work around this, a command is included to run the survey from a development server:
```
Comment (lmorchard) [inline README.md]: This works for me. But, just curious: Have you seen [Storybook](https://getstorybook.io/) for React? IMO, seems like an interesting way to host components in a web page with sample data. Been handy for fiddling with UI & CSS for Snooze Tabs outside the Web Extension context - hot reloading of JS & CSS even!

I'm finding actually hacking on Web Extensions in situ to be kind of fraught. Using the Browser Toolbox for console & DOM inspector & style tweaking, and it seems a bit buggy & prone to disconnection. So, I'm interested in your experience here too

**Review by lmorchard — APPROVED** [review]
Works for me locally, looks like a good start! I think @pdehaan has covered all the nits I might have had. Looks like a minor conflict in the README, but I'd say go ahead and self-merge after optionally addressing any other nits

File: src/webextension/survey/components/app.jsx [hunk 10]
```diff
@@ -0,0 +1,64 @@
+import React from 'react';
```
Comment (pdehaan) [inline src/webextension/survey/components/app.jsx]: Nit: `import React, { Component } from 'react';` for consistency.

File: src/webextension/survey/components/fields/details.jsx [hunk 11]
```diff
@@ -0,0 +1,15 @@
+import React, { Component } from 'react';
+
+export default class DetailsField extends Component {
+  render() {
+    const { input } = this.props;
+    return (
+      <div className="field">
+        <label htmlFor={input.name}>Any other comments?</label>
```
Comment (pdehaan) [inline src/webextension/survey/components/fields/details.jsx]: Nit: Clicking these `<label>` fields didn't really work for me locally unless I modified the DOM in the dev tools to also add `id="foo"` in addition to the existing `name="foo"`.

File: src/webextension/survey/components/fields/reason.jsx [hunk 12]
```diff
@@ -0,0 +1,80 @@
...
+    );
+  }
+
+  render() {
+    const { input, sentiment } = this.props;
+    if (!sentiment) {
+      return null;
+    }
+    return (
+      <div className="field{ this.required }">
```
Comment (pdehaan) [inline src/webextension/survey/components/fields/reason.jsx]: Looks like this `className` may not be interpolating:

pulse_and_builds_user_sentiment_survey__closes__10__closes__18___by_chuckharmston_ _pull_request__82_ _mozilla_pulse_and_1____npm_bin__jpm_run_--addon-dir_build_-b_nightly__firefox-bin__and_index_js_ _build_webextension_survey_ _pulse

**Pull request merged**